### ISSUE TITLE
zsim termination problem for multi-program workloads

### ISSUE THREAD

**Issue by xiaoyaozi5566** [s5z/zsim#101]
I am using zsim with DRAMSim2 for my research. However, I encountered a problem which relates to the termination of multi-program workloads. For example, in a two-core system where each core runs a separate SPEC benchmark, I set each process with the following option:

ffiPoints = "2000000 1000000"

Both programs will fast-forward for 2 million instructions and then simulate for 1 million instructions. However, if program_1 runs faster than program_2, program_1 will stop executing more instructions once it reaches the instruction limit and wait for program_2 to finish. This is not what I want, since in this way program_2 will use the entire DRAM bandwidth (but I want to simulate the contention between the programs) for the rest of the simulation.

I notice there is another option called "maxMinInstrs" which terminates the simulation when all programs reach the instruction limit. This simulation pattern is what I want, but comes with two programs.

1. Does not work with fast-forward (at least I don't know how to)
2. The IPC is reported for all executed instructions, but I only want to know the IPC for the first 1 million instructions for each program.

In summary, my question is, is there a way that terminates all programs only when every program has reached the instruction limit while only reporting the IPC number for the first X number of instructions of each program (of course, should support FF).

Thanks!

**Comment by s5z** [comment]
Yeah, at this point you can combine the multiple ffwd-control flags to do all sorts of things...

In each process, use:

ffiPoints = "2000000 10000000000000";
dumpInstrs = 50000000L;
startFastForwarded = True;

Which fast-forwards 2M instrs, dumps stats into zsim-ev.h5 (the eventual stats) whenever the process reaches 50M instrs, and keeps going. You can have different thresholds per process.

Then, use sim.maxProcEventualDumps = <number of processes you're simulating>. The simulation finishes when all processes have reached their dumpInstrs point. Finally, you can use the stats samples in zsim-ev.h5 to figre out the IPC you want for each process.

**Comment by xiaoyaozi5566** [comment]
Thanks for the help! I have another question regarding sim.maxProcEventualDumps. If program_1 runs much faster than program_2, for example, when program_1 reaches 100M instrs, program_2 haven't reached 50M instrs yet, does that mean both dumps will be for program_1 and the simulation terminates? (program_2 gets no dumps).

**Comment by s5z** [comment]
That won't happen, dumpInstrs is a one-time threshold, not a periodic one.
